src/bin/bench-tps.rs

```diff
@@ -105,35 +105,97 @@ fn sample_tx_count(
+        if confirmatiom.is_ok() {
+            println!("barrier transaction confirmed in {}ms", duration_ms);
+
+            metrics::submit(
+                influxdb::Point::new("bench-tps")
+                    .add_tag(
+                        "op",
+                        influxdb::Value::String("send_barrier_transaction".to_string()),
+                    )
+                    .add_field("poll_count", influxdb::Value::Integer(poll_count))
+                    .add_field("duration", influxdb::Value::Integer(duration_ms as i64))
+                    .to_owned(),
+            );
+
+            // Sanity check that the client balance is still 1
+            let balance = barrier_client.poll_get_balance(&id.pubkey()).unwrap_or(-1);
+            assert_eq!(balance, 1);
+            break;
+        }
+
+        // Timeout after 3 minutes.  When running a CPU-only leader+validator+drone+bench-tps on a dev
+        // machine, some batches of transactions can take upwards of 1 minute...
+        if duration_ms > 1000 * 60 * 3 {
+            println!("Error: Couldn't confirm barrier transaction!");
+            exit(1);
+        }
+
+        let new_last_id = barrier_client.get_last_id();
+        if new_last_id == *last_id {
+            if poll_count > 0 && poll_count % 8 == 0 {
+                println!("last_id is not advancing, still at {:?}", *last_id);
+            }
+        } else {
+            *last_id = new_last_id;
+        }
+
+        poll_count += 1;
+    }
+}
+
 fn generate_txs(
-    client: &mut ThinClient,
     shared_txs: &Arc<RwLock<VecDeque<Vec<Transaction>>>>,
     id: &KeyPair,
     keypairs: &[KeyPair],
-    txs: i64,
-    last_id: &mut Hash,
+    tx_count: i64,
+    last_id: &Hash,
     threads: usize,
     reclaim: bool,
 ) {
-    println!("Signing transactions... {} (reclaim={})", txs / 2, reclaim);
+    println!(
+        "Signing transactions... {} (reclaim={})",
+        tx_count / 2,
+        reclaim
+    );
     let signing_start = Instant::now();
 
-    let transactions: Vec<_> = if !reclaim {
-        keypairs
-            .par_iter()
-            .map(|keypair| Transaction::new(&id, keypair.pubkey(), 1, *last_id))
-            .collect()
-    } else {
-        keypairs
-            .par_iter()
-            .map(|keypair| Transaction::new(keypair, id.pubkey(), 1, *last_id))
-            .collect()
-    };
+    let transactions = keypairs
+        .par_iter()
+        .map(|keypair| {
+            if !reclaim {
+                Transaction::new(&id, keypair.pubkey(), 1, *last_id)
+            } else {
+                Transaction::new(keypair, id.pubkey(), 1, *last_id)
+            }
+        })
+        .collect();
 
     let duration = signing_start.elapsed();
     let ns = duration.as_secs() * 1_000_000_000 + u64::from(duration.subsec_nanos());
-    let bsps = txs as f64 / ns as f64;
-    let nsps = ns as f64 / txs as f64;
+    let bsps = (tx_count / 2) as f64 / ns as f64;
```

Comment:
Looks like we need a ` >> 1` here (and on the next line too) to generate the right stats number

Reply:
Should we just switch to `keypairs.len()` for this function? or `let tx_count = keypairs.len();`

Reply:
Yeah, I'll do that.   All the `/2` is pretty gnarly.